mimic/rest/identity_api.py

```diff
@@ -669,6 +675,518 @@ def list_tenants(self, request):
+                    "Service API for endoint template not found",
+                    request
+                )
+            )
+
+        try:
+            service.add_template(endpoint_template_instance)
+        except ValueError:
+            return json.dumps(
+                conflict(
+                    "Endpoint already exists or service type does not match.",
+                    request
+                )
+            )
+        else:
+            request.setResponseCode(201)
+            return b''
+
+    @app.route('/v2.0/OS-KSCATALOG/endpointTemplates/<string:template_id>',
+               methods=['PUT'])
+    def update_endpoint_templates(self, request, template_id):
+        """
+        Update an API endpoint template already in the system.
+
+        Reference: http://developer.openstack.org/api-ref-identity-v2-ext.html
+
+        Note: A template by the same id must already exist in the system.
+
+        Note: Either the service-id must be specified in the header or
+            a Service Name by the same name must already exist. Otherwise
+            a Not Found (404) will be returned.
+        """
+        x_auth_token = request.getHeader(b"x-auth-token")
+        if x_auth_token is None:
+            return json.dumps(unauthorized("Authentication required", request))
+
+        try:
+            content = json_from_request(request)
+        except ValueError:
+            return json.dumps(
+                bad_request("Invalid JSON request body", request)
+            )
+
+        try:
+            if content['id'] != template_id:
+                return json.dumps(
+                    conflict(
+                        "Template ID in URL does not match that of the JSON body",
+                        request
+                    )
+                )
+
+            endpoint_template_instance = EndpointTemplateStore.deserialize(
+                content
+            )
+        except KeyError as ex:
+            return json.dumps(
+                bad_request(
+                    "JSON body does not contain the required parameters: "
+                    + text_type(ex),
+                    request
+                )
+            )
+
+        service_id = request.getHeader(b'serviceid')
+        if service_id is None:
+            for api_id in self.core.get_external_apis():
+                api = self.core.get_external_api(api_id)
+                if api.has_template(template_id):
+                    service_id = api.uuid_key
+        else:
+            service_id = service_id.decode('utf-8')
+
+        try:
+            service = self.core.get_external_api(service_id)
+        except IndexError:
+            return json.dumps(
+                not_found(
+                    "Service API for endoint template not found",
+                    request
+                )
+            )
+
+        try:
+            service.update_template(endpoint_template_instance)
+        except ValueError:
```

Comment:
I think I should have caught this in an earlier review, but `ValueError` and `IndexError` are bad exceptions to use here.  You can get an `IndexError` or a `ValueError` from basic type issues with built-in Python data structures, and so it's possible to mask basic implementation bugs here and make them very hard to diagnose.  Please fix `update_template` to raise a more specific `EndpointAlreadyExists`, `ServiceTypeMismatch` and `ServiceIdMismatch`.